clippy_lints/src/methods/collapsible_str_replace.rs

```diff
@@ -0,0 +1,160 @@
+use clippy_utils::visitors::for_each_expr;
+use clippy_utils::SpanlessEq;
+use core::ops::ControlFlow;
+use if_chain::if_chain;
+use rustc_ast::ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::{ExprKind, Path, QPath};
+use rustc_lint::LateContext;
+use rustc_span::source_map::Spanned;
+use rustc_span::Span;
+
+use super::method_call;
+use super::COLLAPSIBLE_STR_REPLACE;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>, to: &'tcx hir::Expr<'tcx>) {
+    if_chain! {
+        if let Some(replace_methods) = collect_replace_calls(cx, expr, to);
+        if let Some(earliest_replace_call) = replace_methods.methods.last();
+        if let Some((_, [earliest_recv, ..], _)) = method_call(earliest_replace_call);
+        if cx.typeck_results().expr_ty(earliest_recv).peel_refs().is_str();
+
+        then {
+            // If the parent node's `to` argument is the same as the `to` argument
+            // of the last replace call in the current chain, don't lint as it would be duplicated lint
+            if let Some(parent) = get_parent_expr(cx, expr)
+                && let Some((name, [_, _, current_to], _)) = method_call(parent)
+                && name == "replace"
+                && SpanlessEq::new(cx).eq_expr(to, current_to)
+            {
+                return;
+            }
+
+            // Check if there's an earlier `str::replace` call
+            if let Some((_, [..], span_lo)) = method_call(earliest_replace_call) {
+                check_consecutive_replace_calls(cx, expr, &replace_methods, to, span_lo);
+            }
+        }
+    }
+}
+
+struct ReplaceMethods<'tcx> {
+    methods: Vec<&'tcx hir::Expr<'tcx>>,
+    args: Vec<&'tcx hir::Expr<'tcx>>,
+}
+
+fn collect_replace_calls<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx hir::Expr<'tcx>,
+    to_arg: &'tcx hir::Expr<'tcx>,
+) -> Option<ReplaceMethods<'tcx>> {
+    let mut methods = Vec::new();
+    let mut from_args = Vec::new();
+    let mut return_none = false;
+
+    let _: Option<()> = for_each_expr(expr, |e| {
+        if let Some((name, [_, args @ ..], _)) = method_call(e) {
+            match (name, args) {
+                ("replace", [from, to]) => {
+                    if SpanlessEq::new(cx).eq_expr(to_arg, to)
+                        && cx.typeck_results().expr_ty(from).peel_refs().is_char()
+                    {
+                        methods.push(e);
+                        from_args.push(from);
+                        ControlFlow::Continue(())
+                    } else {
+                        return_none = true;
+                        ControlFlow::BREAK
+                    }
+                },
+                _ => ControlFlow::Continue(()),
+            }
+        } else {
+            ControlFlow::Continue(())
+        }
+    });
+
+    if return_none {
+        return None;
+    }
+
+    Some(ReplaceMethods {
+        methods,
+        args: from_args,
```

Comment:
Why not name this field `from_args`. It seems to be the better name :sweat_smile: